Add prescribed viscosity plugin.

MFraters opened the pull request:
For a project I am working on we need to prescribe a viscosity at a location, which moves with time in a very similar way we can do with the prescribed velocity cookbook plugin. We first hard-coded it in a material model. After a discussion with @jdannberg, @gassmoeller, @alarshi and @RanpengLi, we think this may have broader use. Personally I think this would have been nice to have in previous projects. So we made it into a plugin which can be used as use other material models as a base model. 

The code is basically a combination of the depth dependent plugin and the prescribed velocity plugin.

I do need to finish cleaning the code (see todo's and missing/unupdated documentation and missing changelog entry) and making tests, but I wanted to have this out here early to make sure that this plugin is both wanted and that the naming of the plugin and functions is fine with everyone. I have already tested it for the project I am working on, and it works as intended.

gassmoeller commented on include/aspect/material_model/prescribed_viscosity.h:
```diff
@@ -0,0 +1,109 @@
+         * Parsed function that specifies viscosity in the region described by
+         */
+        Functions::ParsedFunction<dim> prescribed_viscosity_function;
+
```
one empty line

gassmoeller commented on include/aspect/material_model/prescribed_viscosity.h:
```diff
@@ -0,0 +1,109 @@
+     * either through a parsed function, lists of depth and viscosity values, or a file.
+     * The current implementation only allows for depth-dependence of viscosity - all
+     * other properties are derived from the base model.
+     * @ingroup MaterialModels // TODO
```
TODO?

gassmoeller commented on include/aspect/material_model/prescribed_viscosity.h:
```diff
@@ -0,0 +1,109 @@
+#include <deal.II/base/parsed_function.h>
+#include <aspect/material_model/rheology/ascii_depth_profile.h>
+
+#include <boost/lexical_cast.hpp>
```
unnecessary include?

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,206 @@
+*/
+
+#include <aspect/material_model/prescribed_viscosity.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <array>
+
+#include <utility>
+#include <limits>
```
I dont think you need many of these includes?

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,206 @@
+      base_model->evaluate(in,out);
+      for (unsigned int i=0; i<out.n_evaluation_points(); ++i)
+        {
+          if (in.requests_property(MaterialProperties::viscosity))
+            {
+              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);
+              const double viscosity = prescribed_viscosity_function.value(in.position[i]);
+
+              if (indicator > 0.5)
+                {
+                  out.viscosities[i] = viscosity;
+                }
+            }
```
Maybe like this (saves a tiny amount of time):
```suggestion
          if (in.requests_property(MaterialProperties::viscosity))
            {
              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);

              if (indicator > 0.5)
                {
                  out.viscosities[i] = prescribed_viscosity_function.value(in.position[i]);
                }
            }
```

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,206 @@
+
+          prm.declare_entry("Base model","simple",
+                            Patterns::Selection(MaterialModel::get_valid_model_names_pattern<dim>()),
+                            "The name of a material model that will be modified by a depth "
+                            "dependent viscosity. Valid values for this parameter "
+                            "are the names of models that are also valid for the "
+                            "``Material models/Model name'' parameter. See the documentation for "
+                            "that for more information.");
```
Update description.

gassmoeller reviewed (commented):
Yes, looks good and thanks for setting it up. I left some comments.
What we would need in addition:
- a test
- a changelog entry.

MFraters commented:
Thanks for the review! I have addressed your comments and added a test and a changelog.

I made sure that the test is happy, and I hope the tester is as well :)

![happy_viscosity_plugin](https://github.com/geodynamics/aspect/assets/7631629/645575b6-4193-4e04-8215-aa9a144a3e1d)

gassmoeller commented:
/rebuild

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,197 @@
+      }
+      prm.leave_subsection ();
+
+      /* After parsing the parameters for depth dependent, it is essential to parse
```
depth dependent

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,197 @@
+      }
+      prm.leave_subsection ();
+    }
+
```
three empty lines

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,197 @@
+      this->model_dependence = base_model->get_model_dependence();
+    }
+
+    template <int dim>
+    bool
+    PrescribedViscosity<dim>::
+    is_compressible () const
+    {
+      return base_model->is_compressible();
+    }
```
Please move this function above the parse_parameters and declare_parameters functions. This is not a hard rule, but I would like to keep our convention in as many files as possible. Makes it easier to read.

gassmoeller commented on source/material_model/prescribed_viscosity.cc:
```diff
@@ -0,0 +1,197 @@
+                                   "the other available material models. This prescribed viscosity material model "
+                                   "allows the user to specify a function which describes where the viscosity should be "
+                                   "prescribed and a second function which describes the viscosity in that region. "
+                                   "This material model requires a base model which prescribes the viscosity in the "
```
... and the other material parameters ...

gassmoeller commented on tests/happy_prescribed_viscosity_plugin.prm:
maybe rename to `prescribed_viscosity_happy.prm`, this will be easier to find and understand

gassmoeller reviewed (commented):
Thanks for the morale booster :-)
Some comments, but afterwards this should be ready. The failing tests on deal.II master are unrelated.

MFraters commented:
Thanks for the review, I have addressed your comments! :)

gassmoeller reviewed (approved):
Thank you!
